parlai/agents/transformer/polyencoder.py

```diff
@@ -0,0 +1,267 @@
+                           help='number of vectors used to represent the context'
+                                'in the case of n_first, those are the number'
+                                'of vectors that are considered.')
+        agent.add_argument('--poly-attention-type', type=str, default='basic',
+                           choices=['basic', 'basic_sqrt', 'multihead'],
+                           help='Type of the top aggregation layer of the poly-'
+                                'encoder (where the candidate representation is'
+                                'the key)')
+        agent.add_argument('--poly-attention-num-heads', type=int, default=4,
+                           help='In case poly-attention-type is multihead, '
+                                'specify the number of heads')
+
+        # Those arguments are here in case where polyencoder type is 'code'
+        agent.add_argument('--codes-attention-type', type=str, default='basic',
+                           choices=['basic', 'basic_sqrt', 'multihead'],
+                           help='Type ')
+        agent.add_argument('--codes-attention-num-heads', type=int, default=4,
+                           help='In case codes-attention-type is multihead, '
+                                'specify the number of heads')
+        return agent
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        self.rank_loss = torch.nn.CrossEntropyLoss(reduce=True, size_average=True)
+        if self.use_cuda:
+            self.rank_loss.cuda()
+        self.data_parallel = opt.get('data_parallel') and self.use_cuda
+        if self.data_parallel:
+            from parlai.core.distributed_utils import is_distributed
+            if is_distributed():
+                raise ValueError(
+                    'Cannot combine --data-parallel and distributed mode'
+                )
+            self.model = torch.nn.DataParallel(self.model)
+        self.START_IDX = self.END_IDX
+
+
+    def build_model(self, states=None):
+        self.model = PolyEncoderModule(self.opt, self.dict, self.NULL_IDX)
+        return self.model
+
+    def vectorize(self, *args, **kwargs):
+        kwargs['add_start'] = True
+        kwargs['add_end'] = True
+        obs = super().vectorize(*args, **kwargs)
+        return obs
+
+    def _set_text_vec(self, *args, **kwargs):
+        """ Add start and end tokens for set_text_vec()
+        """
+        obs = super()._set_text_vec(*args, **kwargs)
+        if 'text_vec' in obs:
+            obs['text_vec'] = surround(obs['text_vec'],
+                                       self.START_IDX,
+                                       self.END_IDX)
+
+    def score_candidates(self, batch, cand_vecs, cand_encs=None):
+        bsz = batch.text_vec.size(0)
+        ctxt_rep, ctxt_rep_mask, _ = self.model(ctxt_tokens=batch.text_vec)
+
+        if cand_encs is not None:
+            raise Exception('cand_encs is not yet taken into account')
```

Comment:
this should be pretty straight forward to handle, maybe just call `cand_rep = cand_encs`? will have to check that it's the appropriate shape. i added coded to do this in parlai_internal, can send if helpful